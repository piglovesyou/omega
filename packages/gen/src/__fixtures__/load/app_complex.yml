application_id: user
label: User
version: 1
fields:
    - field_id: name
      label: Name
      type: text
      valid_if:
          $required: true

    - field_id: age
      label: Age
      type: number
      shown_if:
          name:
              $required: true
              $gt: 10
              $lte: 100
