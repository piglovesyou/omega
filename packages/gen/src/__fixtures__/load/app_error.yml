application_id: errorschema
label: Error Schema
version: 1
fields:
    - field_id: title
      label: Title
      type: 1
      shown_if:
          x: boom
