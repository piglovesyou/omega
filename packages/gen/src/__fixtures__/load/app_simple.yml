application_id: text
label: Text
version: 1
fields:
    - field_id: title
      label: Title
      type: text
