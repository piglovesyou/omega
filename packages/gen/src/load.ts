import { Application, CondRoot } from '@omega/core';
import { Validator } from 'jsonschema';
import { join } from 'path';
import { env } from 'string-env-interpolation';
import { parse as parseYaml } from 'yaml';
import ApplicationJSONSchema from './__assets__/application.json';
import { readFile } from './lib/file';
import { printError } from './lib/print';

const validator = new Validator();

validator.addSchema(ApplicationJSONSchema as any);

export async function readSchema(fullPath: string) {
  const content = await readFile(fullPath);
  const interpolated = env(content);
  return parseYaml(interpolated);
}

function validateSchema(schema: Application) {
  return validator.validate(schema, { $ref: '#/definitions/Application' });
}

function validateDependees(schema: Application) {
  const errors: any[] = [];
  const table = new Map<string, true>(
    schema.fields.map((e) => [e.field_id, true]),
  );
  function checkIfExists(cond: CondRoot | undefined) {
    if (!cond) return;
    for (const fieldId of Object.keys(cond))
      if (!table.has(fieldId))
        errors.push(new Error(`${fieldId} does not exist`));
  }
  for (const field of schema.fields) {
    checkIfExists(field.shown_if);
  }
  return errors;
}

export async function loadApplicationSchema({
  cwd,
  schemaPath,
}: {
  cwd: string;
  schemaPath: string;
}) {
  const schemaFullPath = join(cwd, schemaPath);
  const schema = await readSchema(schemaFullPath);
  const jsonSchemaValidationResult = validateSchema(schema);
  if (jsonSchemaValidationResult.errors.length) {
    for (const err of jsonSchemaValidationResult.errors) printError(err);
    throw new Error(
      `${schemaPath} has the above error(s).\n\n${jsonSchemaValidationResult.errors
        .map((e) => `- ${e.message}`)
        .join('\n')}`,
    );
  }
  const dependeeErrors = validateDependees(schema);
  if (dependeeErrors.length) {
    for (const err of dependeeErrors) printError(err);
    throw new Error(`field_id in ${schemaPath} points to one does not exist.`);
  }
  return schema as Application;
}
