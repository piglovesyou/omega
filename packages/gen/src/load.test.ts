import { join } from 'path';
import { loadApplicationSchema } from './load';

const cwd = join(__dirname, '__fixtures__/load');

describe('load.ts', () => {
  const orig = console.error;
  beforeEach(() => (console.error = () => {}));
  afterEach(() => (console.error = orig));

  test.each([['app_simple.yml'], ['app_complex.yml']])(
    'loads %p',
    async (schemaPath) => {
      const schema = await loadApplicationSchema({
        cwd,
        schemaPath,
      });
      expect(schema).toMatchSnapshot();
    },
  );

  // test('loads complex application schema', async () => {
  //   const schema = await loadApplicationSchema({
  //     cwd,
  //     schemaPath: 'app_complex.yml',
  //   });
  //   expect(schema).toMatchSnapshot();
  // });

  test('cannot load error schema', async () => {
    const fn = () =>
      loadApplicationSchema({ cwd, schemaPath: 'app_error.yml' });
    await expect(fn).rejects.toThrowErrorMatchingSnapshot();
  });
});
