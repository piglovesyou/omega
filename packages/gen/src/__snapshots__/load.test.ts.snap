// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load.ts cannot load error schema 1`] = `
"app_error.yml has the above error(s).

- is not any of [subschema 0],<#/definitions/HTMLRadioField>,[subschema 2],[subschema 3]"
`;

exports[`load.ts loads "app_complex.yml" 1`] = `
Object {
  "application_id": "user",
  "fields": Array [
    Object {
      "field_id": "name",
      "label": "Name",
      "type": "text",
      "valid_if": Object {
        "$required": true,
      },
    },
    Object {
      "field_id": "age",
      "label": "Age",
      "shown_if": Object {
        "name": Object {
          "$gt": 10,
          "$lte": 100,
          "$required": true,
        },
      },
      "type": "number",
    },
  ],
  "label": "User",
  "version": 1,
}
`;

exports[`load.ts loads "app_simple.yml" 1`] = `
Object {
  "application_id": "text",
  "fields": Array [
    Object {
      "field_id": "title",
      "label": "Title",
      "type": "text",
    },
  ],
  "label": "Text",
  "version": 1,
}
`;
